pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -218,21 +214,13 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 			}
 			// This should be only set for the first request. Guard with ID check regardless.
 			if discReq.Node != nil && discReq.Node.Id != "" {
-				err = s.initConnectionNode(discReq.Node, con)
-				if err != nil {
+				if err = s.initConnection(discReq.Node, con); err != nil {
 					return err
 				}
 			}
 
-			con.mu.Lock()
-			if !con.added {
-				con.added = true
-				con.mu.Unlock()
-				s.addCon(con.ConID, con)
-				defer s.removeCon(con.ConID, con)
-			} else {
-				con.mu.Unlock()
-			}
+			// initConnection would have added the connection, so remove it when it is closed/terminated.
+			defer s.removeCon(con.ConID, con)
```

Comment:
It is a bad specification  to make `defer` separate from the init function, i would suggest return a clanup function from initConnection, like `Context`

Reply:
called each time a request come in?

Reply:
yeah, I was also thinking about separating `defer`  from the init function. Let me refactor that.